File: DSA/tree/binary_search_tree.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Callable, Generic, Iterable, Iterator, Optional, TypeVar

T = TypeVar("T")


@dataclass
class _Node(Generic[T]):
    key: T
    left: Optional["_Node[T]"] = None
    right: Optional["_Node[T]"] = None


class BinarySearchTree(Generic[T]):
    """Unbalanced binary search tree with insert, search, delete, traversal."""

    __slots__ = ("_root", "_size", "_cmp")

    def __init__(self, initial: Optional[Iterable[T]] = None, key: Optional[Callable[[T], T]] = None) -> None:
        self._root: Optional[_Node[T]] = None
        self._size: int = 0
        # key is used to transform values for comparison; default is identity
        self._cmp: Callable[[T], T] = key if key is not None else (lambda x: x)
        if initial is not None:
            for item in initial:
                self.insert(item)

    def __len__(self) -> int:
        return self._size

    def __iter__(self) -> Iterator[T]:
        yield from self.inorder()

    def __repr__(self) -> str:
        return f"BinarySearchTree([{', '.join(repr(x) for x in self)}])"

    def insert(self, value: T) -> None:
        def _insert(node: Optional[_Node[T]], value: T) -> _Node[T]:
            if node is None:
                self._size += 1
                return _Node(value)
            if self._cmp(value) < self._cmp(node.key):
                node.left = _insert(node.left, value)
            elif self._cmp(value) > self._cmp(node.key):
                node.right = _insert(node.right, value)
            # ignore duplicates
            return node

        self._root = _insert(self._root, value)

    def contains(self, value: T) -> bool:
        node = self._root
        while node is not None:
            if self._cmp(value) == self._cmp(node.key):
                return True
            if self._cmp(value) < self._cmp(node.key):
                node = node.left
            else:
                node = node.right
        return False

    def remove(self, value: T) -> bool:
        removed = False

        def _remove(node: Optional[_Node[T]], value: T) -> Optional[_Node[T]]:
            nonlocal removed
            if node is None:
                return None
            if self._cmp(value) < self._cmp(node.key):
                node.left = _remove(node.left, value)
            elif self._cmp(value) > self._cmp(node.key):
                node.right = _remove(node.right, value)
            else:
                removed = True
                if node.left is None:
                    return node.right
                if node.right is None:
                    return node.left
                # replace with inorder successor
                succ_parent = node
                succ = node.right
                while succ.left is not None:
                    succ_parent = succ
                    succ = succ.left
                node.key = succ.key
                if succ_parent.left is succ:
                    succ_parent.left = succ.right
                else:
                    succ_parent.right = succ.right
            return node

        self._root = _remove(self._root, value)
        if removed:
            self._size -= 1
        return removed

    def inorder(self) -> Iterator[T]:
        def _in(node: Optional[_Node[T]]):
            if node is not None:
                yield from _in(node.left)
                yield node.key
                yield from _in(node.right)

        yield from _in(self._root)

    def preorder(self) -> Iterator[T]:
        def _pre(node: Optional[_Node[T]]):
            if node is not None:
                yield node.key
                yield from _pre(node.left)
                yield from _pre(node.right)

        yield from _pre(self._root)

    def postorder(self) -> Iterator[T]:
        def _post(node: Optional[_Node[T]]):
            if node is not None:
                yield from _post(node.left)
                yield from _post(node.right)
                yield node.key

        yield from _post(self._root)

    def clear(self) -> None:
        self._root = None
        self._size = 0

File: DSA/tree/BinarySearchTree.java
package DSA.tree;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.StringJoiner;

/**
 * Unbalanced Binary Search Tree with insert, contains, delete, and traversals.
 */
public final class BinarySearchTree<T> {
    private static final class Node<T> {
        T key;
        Node<T> left;
        Node<T> right;
        Node(T key) { this.key = key; }
    }

    private Node<T> root;
    private int size;
    private final Comparator<? super T> comparator;

    public BinarySearchTree() { this(null); }

    public BinarySearchTree(Comparator<? super T> comparator) {
        this.comparator = comparator;
    }

    public int size() { return size; }
    public boolean isEmpty() { return size == 0; }

    public boolean contains(T value) {
        Node<T> n = root;
        while (n != null) {
            int c = compare(value, n.key);
            if (c == 0) return true;
            n = (c < 0) ? n.left : n.right;
        }
        return false;
    }

    public void insert(T value) {
        root = insertRec(root, value);
    }

    private Node<T> insertRec(Node<T> node, T value) {
        if (node == null) {
            size++;
            return new Node<>(value);
        }
        int c = compare(value, node.key);
        if (c < 0) node.left = insertRec(node.left, value);
        else if (c > 0) node.right = insertRec(node.right, value);
        // ignore duplicates
        return node;
    }

    public boolean remove(T value) {
        int before = size;
        root = removeRec(root, value);
        return size < before;
    }

    private Node<T> removeRec(Node<T> node, T value) {
        if (node == null) return null;
        int c = compare(value, node.key);
        if (c < 0) node.left = removeRec(node.left, value);
        else if (c > 0) node.right = removeRec(node.right, value);
        else {
            // delete this node
            if (node.left == null) { size--; return node.right; }
            if (node.right == null) { size--; return node.left; }
            // replace with inorder successor (smallest in right subtree)
            Node<T> succParent = node;
            Node<T> succ = node.right;
            while (succ.left != null) { succParent = succ; succ = succ.left; }
            node.key = succ.key;
            if (succParent.left == succ) succParent.left = succ.right; else succParent.right = succ.right;
            // size already decremented when actual node removed; mimic by decrement here
            size--;
        }
        return node;
    }

    public List<T> inorder() { List<T> out = new ArrayList<>(); inorderRec(root, out); return out; }
    public List<T> preorder() { List<T> out = new ArrayList<>(); preorderRec(root, out); return out; }
    public List<T> postorder() { List<T> out = new ArrayList<>(); postorderRec(root, out); return out; }

    private void inorderRec(Node<T> n, List<T> out) {
        if (n == null) return;
        inorderRec(n.left, out);
        out.add(n.key);
        inorderRec(n.right, out);
    }

    private void preorderRec(Node<T> n, List<T> out) {
        if (n == null) return;
        out.add(n.key);
        preorderRec(n.left, out);
        preorderRec(n.right, out);
    }

    private void postorderRec(Node<T> n, List<T> out) {
        if (n == null) return;
        postorderRec(n.left, out);
        postorderRec(n.right, out);
        out.add(n.key);
    }

    public void clear() {
        root = null;
        size = 0;
    }

    private int compare(T a, T b) {
        if (comparator != null) return comparator.compare(a, b);
        @SuppressWarnings("unchecked")
        Comparable<? super T> ca = (Comparable<? super T>) a;
        return ca.compareTo(b);
    }

    @Override
    public String toString() {
        StringJoiner j = new StringJoiner(", ", "BinarySearchTree[", "]");
        for (T t : inorder()) j.add(String.valueOf(t));
        return j.toString();
    }
}
